### ISSUE TITLE
Syntax and semantics of `if` statements

### ISSUE THREAD

**Issue by nickrobinson251** [quinnj/Interfaces.jl#25]
Suppose you have an interface like where you declare whether or not a feature is supported like: `support(::MyInterface, ::SomeFeature) = true`

How do you express "you must implement `support(::MyInterface, ::SomeFeature)::Bool`, _and if that returns `true` then you have to implement `cool_feature(::MyInterface)`_"?

Currently our `if` terms are the odd ones out, because they actually get run(!)

For example we currently have example like:
https://github.com/quinnj/Interfaces.jl/blob/a8fd71fccc3aed0caeb7f11e03ce2030469855e8/test/runtests.jl#L12-L13
which actually calls `Base.IteratorSize(MyIterable)` i.e. the method `Base.IteratorSize(::Type{MyIterable})

And our other kind of example is like:
https://github.com/quinnj/Interfaces.jl/blob/a8fd71fccc3aed0caeb7f11e03ce2030469855e8/test/runtests.jl#L82
which similarly calls `Tables.isrowtable(MyTable)` i.e. the method `Tables.isrowtable(::Type{MyTable})`

This leads me to think that we should write "if `supports` then `cool_feature`" as:
```jl
if supports(MyInterface(), SomeFeature())
    cool_feature(::MyInterface)
end
```
which will call `supports(MyInterface(), SomeFeature())` i.e. the method `support(::MyInterface, ::SomeFeature)`

**but** this also states that all objects that satisfy `MyInterface` have to have a no-arg constructor `MyInterface()`, which might not be something the interface actually requires... but the current `@interface` semantics mean we have to provide an `MyInterface` object for the `if` statement (since it's going to actually get called), and so we end up having to try constructing such an object i.e. take a guess at the constructor

I'm inclined to think this is an issue with relying on the `if` statements actually being called (i.e. we have `Base.IteratorSize(Iterable) == Base.HasLength() ` rather than `Base.IteratorSize(::Type{Iterable})::Base.HasLength`, 
presumably because we need some way to express `Tables.isrowtable(MyTable) == true` and not just `Tables.isrowtable(::Type{MyTable})::Bool`) but i can't yet see how to square this circle

(p.s. this example comes from MathOptInterface.jl, which does have this "if `supports(::Optimizer, ::Feature) == true` then ..." interface... although they don't quite have this exact issue because all the Optimizers i'm aware of do have no-arg constructors despite that not being a strict interface requirement)

**Comment by quinnj** [comment]
Yeah, the core problem here goes back to our discussion on static vs. dynamic interfaces. 

> How do you express "you must implement support(::MyInterface, ::SomeFeature)::Bool, and if that returns true then you have to implement cool_feature(::MyInterface)"?

This statement here is necessarily a dynamic, i.e. runtime query that we can't hope to evaluate statically.

In the MathOptInterface.jl case, it seems to me that they should in fact make this a "static" interface requirement, i.e. the required definition should be `supports(::Type{O}, ::Type{F}) where {O <: Optimizer, F <: Feature}` that would then be possible to evaluate statically like our other equivalent conditional interface requirements we have.

I'm going to do some more thinking on the dynamic case tonight, both on the test side of things and maybe for the interface definition as well.......we'll see where this goes.
